File: contrib/python/agentfs_state.py
"""Reference writer for the agentfs agent-state contract, version agentfs/v1.

An agent declares what it is doing by writing one JSON document into its
workspace directory. This module builds that document and publishes it
atomically, so a reader that opens the file at any moment sees a whole document
rather than a partial one.

Vendor this file. It is one module and imports only the standard library.

    from agentfs_state import State, Status

    State(status=Status.RUNNING, task="Index the tree").write("/srv/agents/indexer")

`agentfs schema` prints the JSON Schema this writer emits against, and
`agentfs validate <workspace>` reports what a reader makes of a document.

The writer refuses what the decoder refuses: an unknown status, a negative
ordinal, an empty string where a value is expected, a timestamp with no offset.
A document it writes carries no error-severity diagnostic.
"""

from __future__ import annotations

import enum
import json
import os
import tempfile
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any, Dict, Mapping, Optional, Union

__all__ = ["SCHEMA_VERSION", "STATE_FILE", "Status", "State", "rfc3339", "write_atomic"]

#: The contract version this writer emits.
SCHEMA_VERSION = "agentfs/v1"

#: The document name a reader looks for in a workspace directory.
STATE_FILE = "state.json"

#: The ceiling on a name-shaped member: agent, model, run_id.
MAX_NAME_CHARS = 256

#: The ceiling on a prose member: task, problem.
MAX_TEXT_CHARS = 4096


class Status(str, enum.Enum):
    """The state an agent declares itself to be in.

    The vocabulary is closed. A reader refuses a value outside it rather than
    guessing, so a status that merely contains one of these words, such as
    "not running", declares nothing.
    """

    #: The agent holds work and is progressing.
    RUNNING = "running"
    #: The agent holds no work and is available.
    IDLE = "idle"
    #: The agent holds work it cannot progress without an external input.
    BLOCKED = "blocked"
    #: The agent stopped because it could not continue.
    ERROR = "error"
    #: The agent completed its work.
    DONE = "done"

    def __str__(self) -> str:
        return self.value


@dataclass
class State:
    """One agent state document.

    ``status`` is authoritative and independent of ``problem``: an agent that
    finished after a recovered fault declares ``DONE`` and a problem describing
    it, and one progressing after a transient failure declares ``RUNNING`` and a
    problem. Deriving one from the other makes both unrepresentable.
    """

    #: The state the agent declares itself to be in. Required.
    status: Status
    #: The agent's name. A reader uses the workspace directory name when absent.
    agent: Optional[str] = None
    #: The work in progress.
    task: Optional[str] = None
    #: The position within the task: a non-negative ordinal or a named phase.
    step: Optional[Union[int, str]] = None
    #: The number of steps the task is expected to take.
    steps_total: Optional[int] = None
    #: The model the agent is running against.
    model: Optional[str] = None
    #: The run directory this document belongs to.
    run_id: Optional[str] = None
    #: A fault description, independent of status.
    problem: Optional[str] = None
    #: How often the agent undertakes to rewrite the document. A document older
    #: than this reads as stale.
    heartbeat_seconds: Optional[float] = None
    #: When the agent began the task. Requires a time zone.
    started_at: Optional[datetime] = None
    #: When the agent wrote the document. Requires a time zone.
    updated_at: Optional[datetime] = None
    #: Integrator-defined string metadata.
    labels: Dict[str, str] = field(default_factory=dict)
    #: Members the contract does not define. A reader preserves and ignores
    #: them.
    extra: Dict[str, Any] = field(default_factory=dict)

    def document(self) -> Dict[str, Any]:
        """Return the document as a mapping, in contract order.

        Raises ValueError for a value a reader would refuse.
        """
        status = Status(self.status)
        out: Dict[str, Any] = {"schema": SCHEMA_VERSION, "status": status.value}

        _put(out, "agent", _text(self.agent, "agent", MAX_NAME_CHARS))
        _put(out, "task", _text(self.task, "task", MAX_TEXT_CHARS))
        _put(out, "step", _step(self.step))
        _put(out, "steps_total", _count(self.steps_total, "steps_total"))
        _put(out, "model", _text(self.model, "model", MAX_NAME_CHARS))
        _put(out, "run_id", _text(self.run_id, "run_id", MAX_NAME_CHARS))
        _put(out, "problem", _text(self.problem, "problem", MAX_TEXT_CHARS))
        _put(out, "heartbeat_seconds", _count(self.heartbeat_seconds, "heartbeat_seconds"))
        _put(out, "started_at", rfc3339(self.started_at) if self.started_at else None)
        _put(out, "updated_at", rfc3339(self.updated_at) if self.updated_at else None)
        _put(out, "labels", _labels(self.labels))

        for name, value in self.extra.items():
            if name in out:
                raise ValueError(
                    f"extra member {name!r} collides with a contract member; "
                    "rename it or set the contract member itself"
                )
            out[name] = value
        return out

    def encode(self) -> bytes:
        """Return the document as the UTF-8 bytes to publish."""
        text = json.dumps(self.document(), indent=2, ensure_ascii=False)
        return (text + "\n").encode("utf-8")

    def write(self, directory: str, name: str = STATE_FILE) -> str:
        """Publish the document into ``directory`` and return the path written."""
        target = os.path.join(directory, name)
        write_atomic(target, self.encode())
        return target


def rfc3339(moment: datetime) -> str:
    """Render ``moment`` as an RFC 3339 date-time with its offset.

    A naive datetime is refused: without an offset the text names no instant,
    and a reader on another host cannot place it.
    """
    if moment.tzinfo is None or moment.utcoffset() is None:
        raise ValueError(
            "a timestamp needs a time zone: use datetime.now(timezone.utc), "
            "or attach one with astimezone()"
        )
    text = moment.isoformat(timespec="seconds")
    return text[:-6] + "Z" if text.endswith("+00:00") else text


def write_atomic(path: str, data: bytes) -> None:
    """Publish ``data`` at ``path`` in one step.

    A writer that opens the target and streams into it publishes every
    intermediate state, and a reader that opens the file meanwhile sees bytes
    that are not JSON. Writing the whole document to a temporary file in the
    same directory and renaming it over the target publishes the change
    atomically: a reader observes either the previous document or this one.

    The temporary file shares the target's directory because rename is atomic
    only within one filesystem. The fsync before the rename orders the bytes
    ahead of the name that points at them, so a document that survives a crash
    is a whole document; the fsync on the directory afterwards is what makes
    the name itself survive.
    """
    directory = os.path.dirname(os.path.abspath(path))
    handle, temporary = tempfile.mkstemp(prefix=".agentfs-state-", suffix=".json", dir=directory)
    try:
        with os.fdopen(handle, "wb") as stream:
            stream.write(data)
            stream.flush()
            os.fsync(stream.fileno())
        os.replace(temporary, path)
    except BaseException:
        try:
            os.unlink(temporary)
        except FileNotFoundError:
            pass
        raise
    _fsync_directory(directory)


def _fsync_directory(directory: str) -> None:
    """Flush the directory entry the rename created.

    Platforms that cannot open a directory as a file report the attempt rather
    than the flush, and the rename is already visible to every reader.
    """
    try:
        fd = os.open(directory, os.O_RDONLY)
    except OSError:
        return
    try:
        os.fsync(fd)
    except OSError:
        pass
    finally:
        os.close(fd)


def _put(out: Dict[str, Any], name: str, value: Any) -> None:
    if value is not None:
        out[name] = value


def _text(value: Optional[str], name: str, ceiling: int) -> Optional[str]:
    if value is None:
        return None
    if not isinstance(value, str):
        raise ValueError(f"{name} holds a string, not {type(value).__name__}")
    if value == "":
        raise ValueError(f"{name} is empty; omit it rather than declaring it empty")
    if len(value) > ceiling:
        raise ValueError(
            f"{name} is {len(value)} characters, past the {ceiling} the contract allows; "
            "a state document is a status declaration, not a log"
        )
    return value


def _step(value: Optional[Union[int, str]]) -> Optional[Union[int, str]]:
    if value is None:
        return None
    if isinstance(value, bool):
        raise ValueError("step holds an ordinal or a phase name, not a boolean")
    if isinstance(value, int):
        if value < 0:
            raise ValueError("step is negative; an ordinal counts from zero")
        return value
    if isinstance(value, str):
        return _text(value, "step", MAX_NAME_CHARS)
    raise ValueError(f"step holds an ordinal or a phase name, not {type(value).__name__}")


def _count(value: Optional[float], name: str) -> Optional[float]:
    if value is None:
        return None
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise ValueError(f"{name} holds a number, not {type(value).__name__}")
    if value < 0:
        raise ValueError(f"{name} is negative")
    return value


def _labels(labels: Mapping[str, str]) -> Optional[Dict[str, str]]:
    if not labels:
        return None
    out: Dict[str, str] = {}
    for key, value in labels.items():
        if not isinstance(key, str) or not isinstance(value, str):
            raise ValueError(f"label {key!r} holds a string key and a string value")
        out[key] = value
    return out


def _example(directory: str) -> str:
    """Write one document, declaring every member the contract defines.

    The example carries a ``problem`` while running, because ``status`` is
    authoritative and independent of it: an agent progressing after a transient
    fault declares both.
    """
    from datetime import timezone

    started = datetime.now(timezone.utc)
    state = State(
        status=Status.RUNNING,
        agent="indexer",
        task="Index the workspace tree",
        step=3,
        steps_total=8,
        model="claude-opus-5",
        run_id="2026-04-08T12-40-11Z-1f3c",
        problem="The first two fetches timed out and were retried.",
        heartbeat_seconds=15,
        started_at=started,
        updated_at=started,
        labels={"host": "runner-3", "queue": "batch"},
    )
    return state.write(directory)


if __name__ == "__main__":
    import sys

    print(_example(sys.argv[1] if len(sys.argv) > 1 else "."))

File: contrib/typescript/agentfs-state.ts
/**
 * Reference writer for the agentfs agent-state contract, version agentfs/v1.
 *
 * An agent declares what it is doing by writing one JSON document into its
 * workspace directory. This module builds that document and publishes it
 * atomically, so a reader that opens the file at any moment sees a whole
 * document rather than a partial one.
 *
 * Vendor this file. It is one module and imports only Node's standard library.
 *
 *     import { Status, writeState } from "./agentfs-state.ts";
 *
 *     writeState("/srv/agents/indexer", { status: Status.Running, task: "Index the tree" });
 *
 * `agentfs schema` prints the JSON Schema this writer emits against, and
 * `agentfs validate <workspace>` reports what a reader makes of a document.
 *
 * The writer refuses what the decoder refuses: an unknown status, a negative
 * ordinal, an empty string where a value is expected, a timestamp that names no
 * instant. A document it writes carries no error-severity diagnostic.
 *
 * Every construct here erases to JavaScript, so Node runs the file as it
 * stands: `node agentfs-state.ts <workspace>`.
 */

import { randomUUID } from "node:crypto";
import { closeSync, fsyncSync, openSync, renameSync, unlinkSync, writeFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";

/** The contract version this writer emits. */
export const SCHEMA_VERSION = "agentfs/v1";

/** The document name a reader looks for in a workspace directory. */
export const STATE_FILE = "state.json";

/** The ceiling on a name-shaped member: agent, model, runId. */
export const MAX_NAME_CHARS = 256;

/** The ceiling on a prose member: task, problem. */
export const MAX_TEXT_CHARS = 4096;

/**
 * The state an agent declares itself to be in.
 *
 * The vocabulary is closed. A reader refuses a value outside it rather than
 * guessing, so a status that merely contains one of these words, such as
 * "not running", declares nothing.
 */
export const Status = {
  /** The agent holds work and is progressing. */
  Running: "running",
  /** The agent holds no work and is available. */
  Idle: "idle",
  /** The agent holds work it cannot progress without an external input. */
  Blocked: "blocked",
  /** The agent stopped because it could not continue. */
  Error: "error",
  /** The agent completed its work. */
  Done: "done",
} as const;

/** One of the statuses the contract defines. */
export type Status = (typeof Status)[keyof typeof Status];

/** Every status the contract defines, in contract order. */
export const STATUSES: readonly Status[] = Object.values(Status);

/**
 * One agent state document.
 *
 * Members read in camel case and are written under the contract's names.
 *
 * `status` is authoritative and independent of `problem`: an agent that
 * finished after a recovered fault declares `Done` and a problem describing it,
 * and one progressing after a transient failure declares `Running` and a
 * problem. Deriving one from the other makes both unrepresentable.
 */
export interface State {
  /** The state the agent declares itself to be in. */
  status: Status;
  /** The agent's name. A reader uses the workspace directory name when absent. */
  agent?: string;
  /** The work in progress. */
  task?: string;
  /** The position within the task: a non-negative ordinal or a named phase. */
  step?: number | string;
  /** The number of steps the task is expected to take. */
  stepsTotal?: number;
  /** The model the agent is running against. */
  model?: string;
  /** The run directory this document belongs to. */
  runId?: string;
  /** A fault description, independent of status. */
  problem?: string;
  /**
   * How often the agent undertakes to rewrite the document. A document older
   * than this reads as stale.
   */
  heartbeatSeconds?: number;
  /** When the agent began the task. */
  startedAt?: Date;
  /** When the agent wrote the document. */
  updatedAt?: Date;
  /** Integrator-defined string metadata. */
  labels?: Record<string, string>;
  /** Members the contract does not define. A reader preserves and ignores them. */
  extra?: Record<string, unknown>;
}

/**
 * Build the document as a plain object, in contract order.
 *
 * Throws when a member holds a value a reader would refuse.
 */
export function stateDocument(state: State): Record<string, unknown> {
  if (!STATUSES.includes(state.status)) {
    throw new Error(`status ${JSON.stringify(state.status)} is outside the vocabulary: ${STATUSES.join(", ")}`);
  }

  const out: Record<string, unknown> = { schema: SCHEMA_VERSION, status: state.status };
  put(out, "agent", text(state.agent, "agent", MAX_NAME_CHARS));
  put(out, "task", text(state.task, "task", MAX_TEXT_CHARS));
  put(out, "step", step(state.step));
  put(out, "steps_total", count(state.stepsTotal, "stepsTotal"));
  put(out, "model", text(state.model, "model", MAX_NAME_CHARS));
  put(out, "run_id", text(state.runId, "runId", MAX_NAME_CHARS));
  put(out, "problem", text(state.problem, "problem", MAX_TEXT_CHARS));
  put(out, "heartbeat_seconds", count(state.heartbeatSeconds, "heartbeatSeconds"));
  put(out, "started_at", state.startedAt === undefined ? undefined : rfc3339(state.startedAt));
  put(out, "updated_at", state.updatedAt === undefined ? undefined : rfc3339(state.updatedAt));
  put(out, "labels", labels(state.labels));

  for (const [name, value] of Object.entries(state.extra ?? {})) {
    if (name in out) {
      throw new Error(
        `extra member ${JSON.stringify(name)} collides with a contract member; ` +
          "rename it or set the contract member itself",
      );
    }
    out[name] = value;
  }
  return out;
}

/** Encode the document as the text to publish. */
export function encodeState(state: State): string {
  return `${JSON.stringify(stateDocument(state), null, 2)}\n`;
}

/** Publish the document into `directory` and return the path written. */
export function writeState(directory: string, state: State, name: string = STATE_FILE): string {
  const target = join(directory, name);
  writeAtomic(target, encodeState(state));
  return target;
}

/**
 * Render `moment` as an RFC 3339 date-time with its offset.
 *
 * A date-time without an offset names no instant, and a reader on another host
 * cannot place it, so the rendering always carries one.
 */
export function rfc3339(moment: Date): string {
  if (Number.isNaN(moment.getTime())) {
    throw new Error("a timestamp names an instant: this Date holds none");
  }
  return moment.toISOString().replace(/\.\d{3}Z$/, "Z");
}

/**
 * Publish `data` at `target` in one step.
 *
 * A writer that opens the target and streams into it publishes every
 * intermediate state, and a reader that opens the file meanwhile sees bytes
 * that are not JSON. Writing the whole document to a temporary file in the same
 * directory and renaming it over the target publishes the change atomically: a
 * reader observes either the previous document or this one.
 *
 * The temporary file shares the target's directory because rename is atomic
 * only within one filesystem. The fsync before the rename orders the bytes
 * ahead of the name that points at them, so a document that survives a crash is
 * a whole document; the fsync on the directory afterwards is what makes the name
 * itself survive.
 */
export function writeAtomic(target: string, data: string): void {
  const directory = dirname(resolve(target));
  const temporary = join(directory, `.agentfs-state-${randomUUID()}.json`);

  const handle = openSync(temporary, "wx", 0o644);
  try {
    writeFileSync(handle, data, { encoding: "utf8" });
    fsyncSync(handle);
  } finally {
    closeSync(handle);
  }

  try {
    renameSync(temporary, target);
  } catch (cause) {
    discard(temporary);
    throw cause;
  }
  fsyncDirectory(directory);
}

/**
 * Flush the directory entry the rename created.
 *
 * Platforms that cannot open a directory as a file report the attempt rather
 * than the flush, and the rename is already visible to every reader.
 */
function fsyncDirectory(directory: string): void {
  let handle: number;
  try {
    handle = openSync(directory, "r");
  } catch {
    return;
  }
  try {
    fsyncSync(handle);
  } catch {
    // The name is visible; only its durability across a crash is at stake.
  } finally {
    closeSync(handle);
  }
}

/** Remove a temporary file whose rename did not happen. */
function discard(path: string): void {
  try {
    unlinkSync(path);
  } catch {
    // The temporary file is already gone.
  }
}

function put(out: Record<string, unknown>, name: string, value: unknown): void {
  if (value !== undefined) {
    out[name] = value;
  }
}

function text(value: string | undefined, name: string, ceiling: number): string | undefined {
  if (value === undefined) {
    return undefined;
  }
  if (typeof value !== "string") {
    throw new Error(`${name} holds a string, not ${typeof value}`);
  }
  if (value === "") {
    throw new Error(`${name} is empty; omit it rather than declaring it empty`);
  }
  // A reader counts code points, so the ceiling is measured in them rather
  // than in the UTF-16 units String.length reports: an emoji is one character
  // to the contract and two to length.
  const characters = [...value].length;
  if (characters > ceiling) {
    throw new Error(
      `${name} is ${characters} characters, past the ${ceiling} the contract allows; ` +
        "a state document is a status declaration, not a log",
    );
  }
  return value;
}

function step(value: number | string | undefined): number | string | undefined {
  if (value === undefined) {
    return undefined;
  }
  if (typeof value === "string") {
    return text(value, "step", MAX_NAME_CHARS);
  }
  if (!Number.isInteger(value)) {
    throw new Error("step holds a whole ordinal or a phase name");
  }
  if (value < 0) {
    throw new Error("step is negative; an ordinal counts from zero");
  }
  return value;
}

function count(value: number | undefined, name: string): number | undefined {
  if (value === undefined) {
    return undefined;
  }
  if (typeof value !== "number" || !Number.isFinite(value)) {
    throw new Error(`${name} holds a finite number`);
  }
  if (value < 0) {
    throw new Error(`${name} is negative`);
  }
  return value;
}

function labels(value: Record<string, string> | undefined): Record<string, string> | undefined {
  if (value === undefined || Object.keys(value).length === 0) {
    return undefined;
  }
  for (const [key, held] of Object.entries(value)) {
    if (typeof held !== "string") {
      throw new Error(`label ${JSON.stringify(key)} holds a string value, not ${typeof held}`);
    }
  }
  return { ...value };
}

/**
 * Write one document, declaring every member the contract defines.
 *
 * The example carries a `problem` while running, because `status` is
 * authoritative and independent of it: an agent progressing after a transient
 * fault declares both.
 */
export function example(directory: string): string {
  const started = new Date();
  return writeState(directory, {
    status: Status.Running,
    agent: "indexer",
    task: "Index the workspace tree",
    step: 3,
    stepsTotal: 8,
    model: "claude-opus-5",
    runId: "2026-04-08T12-40-11Z-1f3c",
    problem: "The first two fetches timed out and were retried.",
    heartbeatSeconds: 15,
    startedAt: started,
    updatedAt: started,
    labels: { host: "runner-3", queue: "batch" },
  });
}

// Node puts the entry point in process.argv[1], so comparing it against this
// file's name runs the example only when the file is invoked directly. The
// comparison is on the name rather than on import.meta, which parses under one
// module system and not the other.
const entry = process.argv[1] === undefined ? "" : resolve(process.argv[1]);
if (entry.endsWith("agentfs-state.ts") || entry.endsWith("agentfs-state.js")) {
  console.log(example(process.argv[2] ?? "."));
}
